### ISSUE TITLE
I am unable to find away to scroll to an index in recyclerlstview 

### ISSUE THREAD

**Issue by ajtwgit** [Flipkart/recyclerlistview#181]
I've submitted a [stackoverflow question](https://stackoverflow.com/questions/50556151/raect-redux-how-can-i-change-the-state-appearance-of-a-component-that-is-backgr) for this issue. However, I don't think it's going to be answered anytime soon. 

I've replaced the componentWillReceiveProps in my SO code with: `    
static getDerivedStateFromProps(nextProps, prevState){
        if(nextProps.scrollToSubjectIndex != prevState.initialRenderIndex){
        this.recycRef = React.createRef();
            console.log("refernec  " + this.recyclerRef)
            // this.recycRef.scrollToIndex(nextProps.scrollToSubjectIndex);
            return {initialRenderIndex : nextProps.scrollToSubjectIndex };
        }
        else return null;
    }`

Basically I want to get the index of a touch input from a drawer and then, I want the recyclerlistview to scroll to that index. Adding a ref to the recyclerlistview yields: `TypeError: Cannot add property value, object is not extensible
`

Can anyone give me hand on making this happen?

**Comment by ajtwgit** [comment]
Current code:


import React, { Component } from "react";
import { RecyclerListView, LayoutProvider, DataProvider } from "recyclerlistview";
import { View, Dimensions, Image, AppRegistry } from "react-native";
import {
    Button,
    Text,
    Container,
    Card,
    CardItem,
    Body,
    Content,
    Header,
    Title,
    Left,
    Icon,
    Right,
    List,
  } from "native-base";

import { connect } from 'react-redux';
import ProfileCard from "../ProfileScreen/ProfileCard";
import People from "../../assets/people";

let { height, width } = Dimensions.get('window');

export class NewProPage extends React.Component {

    static getDerivedStateFromProps(nextProps, prevState){
      console.log(nextProps.scrollToSubjectIndex)
      console.log(prevState.initialRenderIndex)

        if(nextProps.scrollToSubjectIndex != prevState.initialRenderIndex){
            return {initialRenderIndex : nextProps.scrollToSubjectIndex };
        }
        else return null;
    }


    constructor(args) {
        super(args);
        const { navigation } = this.props;
        this.state = {
            dataProvider: new DataProvider((r1, r2) => {
                return r1 !== r2
            }).cloneWithRows(People),
            initialRenderIndex: this.props.scrollToSubjectIndex
        };
        this._layoutProvider = new LayoutProvider((i) => {
            return this.state.dataProvider.getDataForIndex(i).type;
        }, (type, dim) => {
            switch (type) {
                case "HOTEL_ITEM":
                    dim.width = width;
                    dim.height = 83;
                    break;
                case "FL_ITEM":
                    dim.width = width;
                    dim.height = 400;
                    break;
                case "HEADER":
                    dim.width = width;
                    dim.height = 300;
                    break;
                default:
                    dim.width = width;
                    dim.height = 0;

            }
        });
        this._renderRow = this._renderRow.bind(this);

    }

    _renderRow(type, data) {
        switch (type) {
            case "FL_ITEM":
                return <ProfileCard data={data} />;
            default:
                return null;
        }
    };

    render() {
        console.log('View from render:     '    +    this.state.initialRenderIndex)
        return (
            <Container>
                <View style={styles.container}>
                <RecyclerListView
                    initialRenderIndex = {this.state.initialRenderIndex}
                        rowRenderer={this._renderRow}
                        dataProvider={this.state.dataProvider}
                        layoutProvider={this._layoutProvider} />
                </View>
            </Container>
        )
    }
}


const mapStateToProps = state => {
    return {
      scrollToSubjectIndex: state.subjectInSCrollView
    }
}

export default connect(mapStateToProps)(NewProPage);

const styles = {
    container: {
        flex: 1
    },
    header: {
        height: 65,
        backgroundColor: 'orange',
        alignItems: "center",
        flexDirection: "row",
        elevation: 4
    },
    headerText: {
        color: 'white',
        fontSize: 18,
        marginLeft: 16,
        paddingBottom: 3
    },
    backIcon: {
        height: 23,
        width: 23,
        marginLeft: 16

    }
}

**Comment by AbdallaMohamed** [comment]
You cannot reference the class instance in static functions. Your static getDerivedStateFromProps shouldn't contain the keyword `this`. You should just componentDidUpdate instead. And move your React.createRef inside the constructor, or better, define it as a class property.

**Comment by ajtwgit** [comment]
@AbdallaMohamed  Thanks. I've added componentDidUpdate but, my ref is still causing that TypeError:

`export class NewProPage extends React.Component {

    componentDidUpdate(prevProps, prevState, snapshot){
        if(prevState.initialRenderIndex != this.state.initialRenderIndex){

        }

    }
    componentWillReceiveProps(nextProps){
        if(nextProps.scrollToSubjectIndex != this.props.scrollToSubjectIndex){
            this.setState({initialRenderIndex:nextProps.scrollToSubjectIndex })
        }
      }

    constructor(args) {
        super(args);
        const { navigation } = this.props;
        this.myRef = React.createRef();
        this.state = {
            dataProvider: new DataProvider((r1, r2) => {
                return r1 !== r2
            }).cloneWithRows(People),
            initialRenderIndex: this.props.scrollToSubjectIndex
        };
        this._layoutProvider = new LayoutProvider((i) => {
            return this.state.dataProvider.getDataForIndex(i).type;
        }, (type, dim) => {
            switch (type) {
                case "HOTEL_ITEM":
                    dim.width = width;
                    dim.height = 83;
                    break;
                case "FL_ITEM":
                    dim.width = width;
                    dim.height = 400;
                    break;
                case "HEADER":
                    dim.width = width;
                    dim.height = 300;
                    break;
                default:
                    dim.width = width;
                    dim.height = 0;

            }
        });
        this._renderRow = this._renderRow.bind(this);

    }

    _renderRow(type, data) {
        switch (type) {
            case "FL_ITEM":
                return <ProfileCard data={data} />;
            default:
                return null;
        }
    };

    render() {
        console.log('View from render:     '    +    this.state.initialRenderIndex)
        return (
            <Container>
                <View style={styles.container}>
                <RecyclerListView
                ref={this.myRef}
                    initialRenderIndex = {this.state.initialRenderIndex}
                        rowRenderer={this._renderRow}
                        dataProvider={this.state.dataProvider}
                        layoutProvider={this._layoutProvider} />
                </View>
            </Container>
        )
    }
}


const mapStateToProps = state => {
    return {
      scrollToSubjectIndex: state.subjectInSCrollView
    }
}

export default connect(mapStateToProps)(NewProPage);

const styles = {
    container: {
        flex: 1
    },
    header: {
        height: 65,
        backgroundColor: 'orange',
        alignItems: "center",
        flexDirection: "row",
        elevation: 4
    },
    headerText: {
        color: 'white',
        fontSize: 18,
        marginLeft: 16,
        paddingBottom: 3
    },
    backIcon: {
        height: 23,
        width: 23,
        marginLeft: 16

    }
}
`

**Comment by ajtwgit** [comment]
@AbdallaMohamed  Seems to work if I use a callback as the ref:

` render: function() {
    return (
      <RecyclerListView
        ref={function(RecyclerListView) {
          if (RecyclerListView != null) {
            RecyclerListView.scrollToIndex(**5**);
          }
        }} />
    );
  },`

I just need to figure out how to submit a new value to the function. I can only get it working by hard-coding so far.

**Comment by ajtwgit** [comment]
Completed code:

Thanks, Adballa.

`export class NewProPage extends React.Component {

    componentDidUpdate(prevProps, prevState, snapshot){
        if(prevState.initialRenderIndex != this.state.initialRenderIndex){
                this.recyclerRef.scrollToIndex(this.state.initialRenderIndex);
        }

    }
    componentWillReceiveProps(nextProps){
        if(nextProps.scrollToSubjectIndex != this.props.scrollToSubjectIndex){
            this.setState({initialRenderIndex:nextProps.scrollToSubjectIndex })
        }
      }

    constructor(args) {
        super(args);
        const { navigation } = this.props;
        this.state = {
            dataProvider: new DataProvider((r1, r2) => {
                return r1 !== r2
            }).cloneWithRows(People),
            initialRenderIndex: this.props.scrollToSubjectIndex
        };
        this._layoutProvider = new LayoutProvider((i) => {
            return this.state.dataProvider.getDataForIndex(i).type;
        }, (type, dim) => {
            switch (type) {
                case "HOTEL_ITEM":
                    dim.width = width;
                    dim.height = 83;
                    break;
                case "FL_ITEM":
                    dim.width = width;
                    dim.height = 400;
                    break;
                case "HEADER":
                    dim.width = width;
                    dim.height = 300;
                    break;
                default:
                    dim.width = width;
                    dim.height = 0;

            }
        });
        this._renderRow = this._renderRow.bind(this);

    }

    _renderRow(type, data) {
        switch (type) {
            case "FL_ITEM":
                return <ProfileCard data={data} />;
            default:
                return null;
        }
    };

    render() {
        console.log('View from render:     '    +    this.state.initialRenderIndex)
        return (
            <Container>
                <View style={styles.container}>
                <RecyclerListView
                     ref={(RecyclerListView) =>{
                         this.recyclerRef = RecyclerListView;
                    }}
                    initialRenderIndex = {this.state.initialRenderIndex}
                        rowRenderer={this._renderRow}
                        dataProvider={this.state.dataProvider}
                        layoutProvider={this._layoutProvider} />
                </View>
            </Container>
        )
    }
}


const mapStateToProps = state => {
    return {
      scrollToSubjectIndex: state.subjectInSCrollView
    }
}

export default connect(mapStateToProps)(NewProPage);`

**Comment by naqvitalha** [comment]
Great that you got it working. I'll close it, thanks for resolving this @AbdallaMohamed
